crates/packaging/src/tarball.rs

```diff
@@ -269,9 +274,51 @@ fn read_header<'a>(
     // (We can't use that for the parser state and still return Module<'a> unfortunately.)
     let arena_buf = bumpalo::collections::Vec::from_iter_in(buf.iter().copied(), arena);
     let parse_state = State::new(arena_buf.into_bump_slice());
-    let (module, _) = parse_header(arena, parse_state).unwrap_or_else(|_err| {
+    let (module, state) = parse_header(arena, parse_state).unwrap_or_else(|_err| {
         todo!(); // TODO report a nice error and exit 1 - or maybe just return Err, for better testability?
     });
 
-    Ok(module)
+    Ok((module, state))
+}
+
+fn add_ingested_files<W: Write>(
+    arena: &Bump,
+    root_dir: &Path,
+    dot_roc_path: &Path,
+    builder: &mut tar::Builder<W>,
+) -> io::Result<()> {
+    let mut buf = Vec::new();
+    let (module, state) = read_header(arena, &mut buf, dot_roc_path)?;
+    let (_, defs) = module.upgrade_header_imports(arena);
+
+    match parse_module_defs(arena, state, defs) {
+        Ok(defs) => RecursiveValueDefIter::new(&defs).try_for_each(|(def, _)| {
+            if let ValueDef::IngestedFileImport(IngestedFileImport { path, .. }) = def {
+                if let StrLiteral::PlainLine(relative_path) = path.value {
+                    let mut abs_path: PathBuf = relative_path.into();
+                    abs_path.pop();
+                    abs_path.push(relative_path);
+
+                    match abs_path.strip_prefix(root_dir) {
+                        Ok(name) => builder.append_path_with_name(abs_path.as_path(), name),
+                        Err(_) => {
+                            panic!(
+                                "Cannot bundle {} (imported in {}) since it's outside {}",
+                                abs_path.display(),
+                                dot_roc_path.display(),
+                                root_dir.display()
+                            );
```

Comment:
We should use reporting instead of panicking here, but that's not setup for this module yet, I'll do that in a separate PR.